src/main/java/com/ibm/stocator/fs/ObjectStoreFileSystem.java

```diff
@@ -307,13 +308,17 @@ public boolean delete(Path f, boolean recursive) throws IOException {
 
     if ((fileStatus != null && fileStatus.isDirectory()) || (fileStatus == null && prefixBased)) {
       LOG.trace("{} is directory, prefix based listing set to {}", f.toString(), prefixBased);
-      result = storageClient.list(hostNameScheme, f, false, prefixBased);
+      FileStatus[] listing = storageClient.list(hostNameScheme, f, false, prefixBased);
+      for (FileStatus fs : listing) {
+        if (filter.accept(fs.getPath())) {
+          result.add(fs);
+        }
+      }
     } else if (fileStatus != null) {
```

Comment:
@djalova I think if no filter provided, then do return listStatus(f, null); instead of return listStatus(f, DEFAULT_FILTER); . And then do the looop, only if filter!=null.

Yes, additional loop is costly. Better be avoided.